test/test_linalg.py

```diff
@@ -2064,6 +2064,105 @@ def run_test_singular_input(batch_dim, n):
+            self.assertEqual(ans, out)
+            self.assertEqual(ans, A_pinv)
+
+        def run_test_numpy(A, hermitian):
+            # Check against NumPy output
+            # Test float rcond, and specific value for each matrix
+            rconds = [float(torch.rand(1)), torch.rand(A.shape[:-2], dtype=torch.double, device=device)]
+            # Test broadcasting of rcond
+            if A.ndim > 2:
+                rconds.append(torch.rand(A.shape[-3], device=device))
+            for rcond in rconds:
+                actual = torch.linalg.pinv(A, rcond=rcond, hermitian=hermitian)
+                numpy_rcond = rcond if isinstance(rcond, float) else rcond.cpu().numpy()
+                expected = np.linalg.pinv(A.cpu().numpy(), rcond=numpy_rcond, hermitian=hermitian)
+                self.assertEqual(actual, expected)
+
+        for sizes in [(5, 5), (3, 5, 5), (3, 2, 5, 5),  # square matrices
+                      (3, 2), (5, 3, 2), (2, 5, 3, 2),  # fat matrices
+                      (2, 3), (5, 2, 3), (2, 5, 2, 3),  # thin matrices
+                      (0, 0), (0, 2), (2, 0), (3, 0, 0), (0, 3, 0), (0, 0, 3)]:  # zero numel matrices
+            A = torch.randn(*sizes, dtype=dtype, device=device)
+            hermitian = False
+            run_test_main(A, hermitian)
+            run_test_numpy(A, hermitian)
+
+        # Check hermitian = True
+        for sizes in [(5, 5), (3, 5, 5), (3, 2, 5, 5),  # square matrices
+                      (0, 0), (3, 0, 0), ]:  # zero numel square matrices
+            A = random_hermitian_pd_matrix(sizes[-1], *sizes[:-2], dtype=dtype, device=device)
+            hermitian = True
+            run_test_main(A, hermitian)
+            run_test_numpy(A, hermitian)
+
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float64)
+    def test_pinv_autograd(self, device, dtype):
+        from torch.testing._internal.common_utils import random_fullrank_matrix_distinct_singular_value
+
+        n = 5
+        for batches in ([], [2], [2, 3]):
+            # using .to(device) instead of device=device because @xwang233 claims it's faster
+            a = random_fullrank_matrix_distinct_singular_value(n, *batches, dtype=dtype).to(device)
+            a.requires_grad_()
+
+            def func(a, hermitian):
+                if hermitian:
+                    a = a + a.conj().transpose(-2, -1)
+                return torch.linalg.pinv(a, hermitian=hermitian)
+
+            for hermitian in [False, True]:
+                gradcheck(func, [a, hermitian])
+                gradgradcheck(func, [a, hermitian])
+
+    # TODO: RuntimeError: svd does not support automatic differentiation for outputs with complex dtype.
+    # See https://github.com/pytorch/pytorch/pull/47761
+    @unittest.expectedFailure
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.complex128)
+    def test_pinv_autograd_complex_xfailed(self, device, dtype):
```

Comment:
This failure test will be needed.

There are other failure cases that need to be tested for, too. Like rcond having the wrong dtype or the sizes of the inputs being incorrect.

Reply:
I added `test_pinv_errors_and_warnings ` https://github.com/pytorch/pytorch/pull/48399/commits/080064196fa47d5d9c43089ca8064f50e6d4d70a